### PR TITLE
Feat/bingo game indexer

### PR DESCRIPTION
**Pull request by xiboXD**

### CODE REVIEW

File: src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+using BingoGame.Indexer.CA.Processors;
+
+namespace BingoGame.Indexer.CA.Processors;
+
+public class BingoedProcessor : BingoGameProcessorBase<Bingoed>
+{   
+
+    private readonly IAElfIndexerClientEntityRepository<BingoGameIndex, TransactionInfo> _bingoIndexRepository;
+    private readonly IAElfIndexerClientEntityRepository<BingoGameStaticsIndex, TransactionInfo> _bingoStaticsIndexRepository;
+    private readonly IObjectMapper _objectMapper;
```
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs]: There's already one field in the base class. We should use that one.

File: src/BingoGame.Indexer.CA/Processors/PlayedProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+using BingoGame.Indexer.CA.GraphQL;
+using Volo.Abp.ObjectMapping;
+
+namespace BingoGame.Indexer.CA.Processors;
+
+public class PlayedProcessor : BingoGameProcessorBase<Played>
+{
+    private readonly IAElfIndexerClientEntityRepository<BingoGameIndex, TransactionInfo> _bingoIndexRepository;
+    private readonly IAElfIndexerClientEntityRepository<BingoGameStaticsIndex, TransactionInfo> _bingoStaticsIndexRepository;
+    private readonly IObjectMapper _objectMapper;
```
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/Processors/PlayedProcessor.cs]: There's already one field in the base class. We should use that one.

File: src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,108 @@
+using AElfIndexer.Client;
+using AElfIndexer.Client.Handlers;
+using AElfIndexer.Grains.State.Client;
+using Google.Protobuf.WellKnownTypes;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Portkey.Contracts.BingoGameContract;
+using BingoGame.Indexer.CA.Entities;
+using BingoGame.Indexer.CA.GraphQL;
```
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs]: Please remove all un-used namespaces.

File: test/BingoGame.Indexer.CA.Tests/Generated/Acs0.c.cs
Comment (stevenportkey) [inline test/BingoGame.Indexer.CA.Tests/Generated/Acs0.c.cs:1]: Not used. Please remove.

File: test/BingoGame.Indexer.CA.Tests/Generated/Acs0.g.cs
Comment (stevenportkey) [inline test/BingoGame.Indexer.CA.Tests/Generated/Acs0.g.cs:1]: Not used. Please remove.

File: test/BingoGame.Indexer.CA.Tests/Generated/CaContract.c.cs
Comment (stevenportkey) [inline test/BingoGame.Indexer.CA.Tests/Generated/CaContract.c.cs:1]: Not used. Please remove.

File: test/BingoGame.Indexer.CA.Tests/Generated/CaContract.g.cs
Comment (stevenportkey) [inline test/BingoGame.Indexer.CA.Tests/Generated/CaContract.g.cs:1]: Not used. Please remove.

File: src/BingoGame.Indexer.CA/Generated/CaContract.c.cs
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/Generated/CaContract.c.cs:1]: Not used. Please remove.

File: src/BingoGame.Indexer.CA/Generated/CaContract.g.cs
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/Generated/CaContract.g.cs:1]: Not used. Please remove.

File: src/BingoGame.Indexer.CA/Generated/Acs0.c.cs
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/Generated/Acs0.c.cs:1]: Not used. Please remove.

File: src/BingoGame.Indexer.CA/Generated/Acs0.g.cs
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/Generated/Acs0.g.cs:1]: Not used. Please remove.

**Review by stevenportkey — CHANGES_REQUESTED** [review]
Reviewing. Submitting comments so far.

File: src/BingoGame.Indexer.CA/Handlers/BingoGameHandler.cs
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/Handlers/BingoGameHandler.cs:1]: Why do we need this? It's not doing anything.

File: src/BingoGame.Indexer.CA/Processors/BingoGameProcessorBase.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+                     new Dictionary<string, long>();
+        }
+
+        if (extraProperties.TryGetValue("ResourceFee", out var resourceFee))
+        {
+            var resourceFeeMap = JsonConvert.DeserializeObject<Dictionary<string, long>>(resourceFee) ??
+                                 new Dictionary<string, long>();
+            foreach (var (symbol, fee) in resourceFeeMap)
+            {
+                if (feeMap.TryGetValue(symbol, out _))
```
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/Processors/BingoGameProcessorBase.cs]: Use `ContainsKey` to check if the dictionary contains the key.

File: src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        if (eventValue.PlayerAddress == null || eventValue.PlayerAddress.Value == null)
+        {
+            return;
+        }
+        //update bingoIndex
+        var index = await _bingoIndexRepository.GetFromBlockStateSetAsync(eventValue.PlayId.ToHex(), context.ChainId);
+        if (index == null)
+        {
+            return;
+        }
```
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs]: Should we throw exception if the index doesn't exist?

File: src/BingoGame.Indexer.CA/Processors/PlayedProcessor.cs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        if (eventValue.PlayerAddress == null || eventValue.PlayerAddress.Value == null)
+        {
+            return;
+        }
+
+        var index = await _bingoIndexRepository.GetFromBlockStateSetAsync(eventValue.PlayId.ToHex(), context.ChainId);
+        if (index != null)
+        {
+            return;
+        }
```
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/Processors/PlayedProcessor.cs]: Same as above should we throw exception if the index doesn't exist?

File: test/BingoGame.Indexer.CA.Tests/Processors/BingoedLogEventProcessorTests.cs
Comment (stevenportkey) [inline test/BingoGame.Indexer.CA.Tests/Processors/BingoedLogEventProcessorTests.cs:1]: File name doesn't align with source file name.

File: test/BingoGame.Indexer.CA.Tests/Processors/PlayedLogEventProcessorTests.cs
Comment (stevenportkey) [inline test/BingoGame.Indexer.CA.Tests/Processors/PlayedLogEventProcessorTests.cs:1]: File name doesn't align with source file name.

File: src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        {
+            feeList = new List<TransactionFee>
+            {
+                new ()
+                {
+                    Symbol = null,
+                    Amount = 0
+                }
+            };
+        }
```
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs]: Why can't we simply return an empty list?

**Review by stevenportkey — CHANGES_REQUESTED** [review]
We should add test case for negative award.

File: src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs (hunk 1 above)
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs]: noted, will delete

File: src/BingoGame.Indexer.CA/Processors/PlayedProcessor.cs (hunk 2 above)
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/Processors/PlayedProcessor.cs]: noted, will delete

File: src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs (hunk 3 above)
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs]: ok，will remove un-used namespaces

File: test/BingoGame.Indexer.CA.Tests/Generated/Acs0.c.cs
Comment (xiboXD) [reply test/BingoGame.Indexer.CA.Tests/Generated/Acs0.c.cs:1]: noted, will remove this file

File: test/BingoGame.Indexer.CA.Tests/Processors/PlayedLogEventProcessorTests.cs [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        var bingoedLogEventProcessor = GetRequiredService<PlayedProcessor>();
+        
+        await bingoedLogEventProcessor.HandleEventAsync(logEventInfo, logEventContext);
+        
+        bingoedLogEventProcessor.GetContractAddress(chainId);
+        
+        //step4: save blockStateSet into es
+        await BlockStateSetSaveDataAsync<LogEventInfo>(blockStateSetKey);
+        await BlockStateSetSaveDataAsync<TransactionInfo>(blockStateSetKeyTransaction);
+        await Task.Delay(2000);
```
Comment (stevenportkey) [inline test/BingoGame.Indexer.CA.Tests/Processors/PlayedLogEventProcessorTests.cs]: Why do we need to do it here? What happens if we don't do it? How is it done in production?

File: src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+        index.IsComplete = true;
+        index.Dices = eventValue.Dices.Dices.ToList();
+        index.Award = eventValue.Award;
+        index.BingoBlockHash = context.BlockHash;
+        _objectMapper.Map<LogEventContext, BingoGameIndex>(context, index);
+        await _bingoIndexRepository.AddOrUpdateAsync(index);
+        
+        //update bingoStaticsIndex
+        var staticsId= IdGenerateHelper.GetId(context.ChainId, eventValue.PlayerAddress.ToBase58());
+        var bingoStaticsIndex = await _bingoStaticsIndexRepository.GetFromBlockStateSetAsync(staticsId, context.ChainId);
```
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs]: Why don't we use `GetAsync`?

File: test/BingoGame.Indexer.CA.Tests/Generated/Acs0.g.cs
Comment (xiboXD) [reply test/BingoGame.Indexer.CA.Tests/Generated/Acs0.g.cs:1]: noted, will remove this file

File: test/BingoGame.Indexer.CA.Tests/Generated/CaContract.c.cs
Comment (xiboXD) [reply test/BingoGame.Indexer.CA.Tests/Generated/CaContract.c.cs:1]: noted, will remove this file

File: test/BingoGame.Indexer.CA.Tests/Generated/CaContract.g.cs
Comment (xiboXD) [reply test/BingoGame.Indexer.CA.Tests/Generated/CaContract.g.cs:1]: noted, will remove this file

File: src/BingoGame.Indexer.CA/Generated/CaContract.c.cs
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/Generated/CaContract.c.cs:1]: noted, will remove this file

File: src/BingoGame.Indexer.CA/Generated/CaContract.g.cs
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/Generated/CaContract.g.cs:1]: noted, will remove this file

File: src/BingoGame.Indexer.CA/Generated/Acs0.c.cs
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/Generated/Acs0.c.cs:1]: noted, will remove this file

File: src/BingoGame.Indexer.CA/Generated/Acs0.g.cs
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/Generated/Acs0.g.cs:1]: noted, will remove this file

File: src/BingoGame.Indexer.CA/Handlers/BingoGameHandler.cs
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/Handlers/BingoGameHandler.cs:1]: To handle transaction, we need to create this to inherit TransactionDataHandler, otherwise ProcessTransactionsAsync can not be implemented. It will case a block to be scanned multiple times.

File: src/BingoGame.Indexer.CA/Processors/BingoGameProcessorBase.cs (hunk 4 above)
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/Processors/BingoGameProcessorBase.cs]: noted, changed to ContainsKey

File: src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs (hunk 5 above)
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs]: noted, will add

File: src/BingoGame.Indexer.CA/Processors/PlayedProcessor.cs (hunk 6 above)
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/Processors/PlayedProcessor.cs]: noted, added

File: test/BingoGame.Indexer.CA.Tests/Processors/BingoedLogEventProcessorTests.cs
Comment (xiboXD) [reply test/BingoGame.Indexer.CA.Tests/Processors/BingoedLogEventProcessorTests.cs:1]: noted, will change to BingoedProcessorTests.cs

File: test/BingoGame.Indexer.CA.Tests/Processors/PlayedLogEventProcessorTests.cs
Comment (xiboXD) [reply test/BingoGame.Indexer.CA.Tests/Processors/PlayedLogEventProcessorTests.cs:1]: noted, will change to PlayedProcessorTests.cs

File: src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs (hunk 7 above)
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs]: before we returned an empty list but PM said it's better we could return 0 if there is no transaction fee

File: test/BingoGame.Indexer.CA.Tests/Processors/PlayedLogEventProcessorTests.cs (hunk 8 above)
Comment (xiboXD) [reply test/BingoGame.Indexer.CA.Tests/Processors/PlayedLogEventProcessorTests.cs]: this delay is to make sure data being stored in local ES, then we can verify if data in ES is correct, e.g. bingoGameIndexData.Amount.ShouldBe(100000000); in live env, we don't need to verify data in ES, this verification only happens in test cases

File: src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs (hunk 9 above)
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs]: yes, we can use GetAsync

File: src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs (hunk 7 above)
Comment (stevenportkey) [reply src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs]: Add a comment to record this requirement.

File: src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    {
+        if (eventValue.PlayerAddress == null || eventValue.PlayerAddress.Value == null)
+        {
+            return;
+        }
+        //update bingoIndex
+        var index = await _bingoIndexRepository.GetFromBlockStateSetAsync(eventValue.PlayId.ToHex(), context.ChainId);
+        if (index == null)
+        {
+            throw new Exception("Bingo index not found.");
```
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs]: Define a custom exception type.

File: src/BingoGame.Indexer.CA/Processors/PlayedProcessor.cs [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    {   
+        if (eventValue.PlayerAddress == null || eventValue.PlayerAddress.Value == null)
+        {
+            return;
+        }
+
+        var index = await _bingoIndexRepository.GetFromBlockStateSetAsync(eventValue.PlayId.ToHex(), context.ChainId);
+        if (index != null)
+        {
+            throw new Exception("This event is already handled.");
```
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/Processors/PlayedProcessor.cs]: Define a custom exception type.

File: src/BingoGame.Indexer.CA/Entities/BingoGameIndex.cs [hunk 12]
```diff
@@ -0,0 +1,26 @@
+using AElf.Indexing.Elasticsearch;
+using AElfIndexer.Client;
+using Nest;
+using BingoGame.Indexer.CA.GraphQL;
+
+namespace BingoGame.Indexer.CA.Entities;
+
+public class BingoGameIndex : AElfIndexerClientEntity<string>, IIndexBuild
```
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/Entities/BingoGameIndex.cs]: The name is wrong. It's not an index. Let's call it `BingoGameIndexEntry`.

File: src/BingoGame.Indexer.CA/Entities/Guardian.cs [hunk 13]
```diff
@@ -0,0 +1,15 @@
+using Nest;
+
+namespace BingoGame.Indexer.CA.Entities;
+
+public class Guardian
```
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/Entities/Guardian.cs]: It's not used anywhere. We should delete it.

File: src/BingoGame.Indexer.CA/GraphQL/BingoResultDto.cs [hunk 14]
```diff
@@ -0,0 +1,50 @@
+using AElf.Indexing.Elasticsearch;
+using AElfIndexer.Client;
+using GraphQL;
+using Nest;
+using BingoGame.Indexer.CA.GraphQL;
+
+namespace BingoGame.Indexer.CA.Entities;
```
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/GraphQL/BingoResultDto.cs]: The namespace doesn't align with the folder name. Please fix it.

File: src/BingoGame.Indexer.CA/GraphQL/GetBingoDto.cs [hunk 15]
```diff
@@ -0,0 +1,15 @@
+using GraphQL;
+using Nest;
+using BingoGame.Indexer.CA.GraphQL;
+using Volo.Abp.Application.Dtos;
+
+namespace BingoGame.Indexer.CA.Entities;
```
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/GraphQL/GetBingoDto.cs]: The namespace doesn't align with the folder name. Please fix it.

**Review by stevenportkey — CHANGES_REQUESTED** [review]

File: src/BingoGame.Indexer.CA/GraphQL/Query.cs [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+    {
+        var mustQuery = new List<Func<QueryContainerDescriptor<BingoGameIndex>, QueryContainer>>();
+
+        mustQuery.Add(q => q.Term(i => i.Field(f => f.PlayId).Value(dto.PlayId)));
+        mustQuery.Add(q => q.Term(i => i.Field(f => f.IsComplete).Value(true)));
+        // mustQuery.Add(q => q.Term(i => i.Field(f => f.player_address).Value(dto.CAAddressInfos)));
+        // mustQuery.Add(q => q.Term(i => i.Field(f => f.TokenInfo.Symbol).Value(dto.Symbol)));
+        // mustQuery.Add(q => q.Term(i => i.Field(f => f.TokenInfo.Type).Value(dto.Type)));
+
+        // mustQuery.Add(q=> q.Range(i => i.Field(f => f.Balance).GreaterThan(0)));
```
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/GraphQL/Query.cs]: Please remove un-used code.

File: src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs (hunk 10 above)
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/Processors/BingoedProcessor.cs]: noted, added BingoGameIndexEntryNotFoundException

File: src/BingoGame.Indexer.CA/Processors/PlayedProcessor.cs (hunk 11 above)
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/Processors/PlayedProcessor.cs]: noted, added PlayEventAlreadyHandledException

File: src/BingoGame.Indexer.CA/Entities/BingoGameIndex.cs (hunk 12 above)
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/Entities/BingoGameIndex.cs]: noted, changed all BingoGameIndex in the project to BingoGameIndexEntry

File: src/BingoGame.Indexer.CA/Entities/Guardian.cs (hunk 13 above)
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/Entities/Guardian.cs]: noted, deleted

File: src/BingoGame.Indexer.CA/GraphQL/BingoResultDto.cs (hunk 14 above)
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/GraphQL/BingoResultDto.cs]: noted, changed Entities to GraphQL

File: src/BingoGame.Indexer.CA/GraphQL/GetBingoDto.cs (hunk 15 above)
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/GraphQL/GetBingoDto.cs]: noted, changed Entities to GraphQL

File: src/BingoGame.Indexer.CA/GraphQL/Query.cs (hunk 16 above)
Comment (xiboXD) [reply src/BingoGame.Indexer.CA/GraphQL/Query.cs]: noted, deleted all un-used code.

File: src/BingoGame.Indexer.CA/GraphQL/Query.cs [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+            {
+                var mustQueryAddressInfo = new List<Func<QueryContainerDescriptor<BingoGameIndexEntry>, QueryContainer>>
+                {
+                    q => q.Term(i => i.Field(f => f.PlayerAddress).Value(address))
+                };
+                shouldQuery.Add(q => q.Bool(b => b.Must(mustQueryAddressInfo)));
+            }
+
+            mustQuery.Add(q => q.Bool(b => b.Should(shouldQuery)));
+        }
```
Comment (stevenportkey) [inline src/BingoGame.Indexer.CA/GraphQL/Query.cs]: Replace is with `Terms`.

Comment (xiboXD) [reply src/BingoGame.Indexer.CA/GraphQL/Query.cs]: noted, changed to terms query

**Review by kevin00k — APPROVED** [review]

**Review by gldeng — APPROVED** [review]

**Review by gldeng — APPROVED** [review]

**Review by stevenportkey — APPROVED** [review]
